ecs-agent/wsclient/client_test.go

```diff
@@ -0,0 +1,424 @@
+
+	ctx := context.Background()
+	mockServer, _, _, _, _ := utils.GetMockServer(closeWS)
+	mockServer.StartTLS()
+
+	types := []interface{}{ecsacs.PayloadMessage{}}
+	messageError := make(chan error)
+	cs := getTestClientServer(mockServer.URL, types, 1)
+	cs.conn = conn
+
+	defer cs.Close()
+
+	messageChannel := make(chan *ecsacs.PayloadMessage)
+	reqHandler := func(payload *ecsacs.PayloadMessage) {
+		messageChannel <- payload
+	}
+	cs.AddRequestHandler(reqHandler)
+
+	go func() {
+		messageError <- cs.ConsumeMessages(ctx)
+		cs.Close()
+	}()
+
+	expectedMessage := &ecsacs.PayloadMessage{
+		Tasks: []*ecsacs.Task{{
+			Arn: aws.String("arn"),
+		}},
+	}
+
+	assert.Equal(t, expectedMessage, <-messageChannel)
+
+}
+
+// TestMakeUnrecognizedRequest tests if the correct error type is returned
+// on unrecognized request type.
+func TestMakeUnrecognizedRequest(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	conn := mock_wsconn.NewMockWebsocketConn(ctrl)
+	conn.EXPECT().SetWriteDeadline(gomock.Any()).Return(nil)
+	conn.EXPECT().Close()
+
+	closeWS := make(chan []byte)
+	defer close(closeWS)
+
+	mockServer, _, _, _, _ := utils.GetMockServer(closeWS)
+	mockServer.StartTLS()
+
+	types := []interface{}{ecsacs.PayloadMessage{}}
+	cs := getTestClientServer(mockServer.URL, types, 1)
+	cs.conn = conn
+
+	defer cs.Close()
+
+	err := cs.MakeRequest(t)
+	if _, ok := err.(*UnrecognizedWSRequestType); !ok {
+		t.Fatal("Expected unrecognized request type")
+	}
+}
+
+// TestWriteCloseMessage tests if the wsclient can successfully close the connection
+// and write close message. The close message is expected to be received on server side.
+func TestWriteCloseMessage(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	closeWS := make(chan []byte)
+	defer close(closeWS)
+
+	mockServer, _, _, errChan, _ := utils.GetMockServer(closeWS)
+	mockServer.StartTLS()
+
+	types := []interface{}{ecsacs.PayloadMessage{}}
+	cs := getTestClientServer(mockServer.URL, types, 1)
+	cs.Connect()
+
+	defer cs.Close()
+
+	err := cs.WriteCloseMessage()
+	assert.NoError(t, err)
+	assert.Error(t, <-errChan)
+}
+
+// TestCtxCancel tests if the passed context, on receiving the cancel
+// on the created ctx.Done channel, performs the expected behavior of
+// closing the connection and returns the ctx error.
+func TestCtxCancel(t *testing.T) {
+	closeWS := make(chan []byte)
+	defer close(closeWS)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	messageError := make(chan error)
+	mockServer, _, _, _, _ := utils.GetMockServer(closeWS)
+	mockServer.StartTLS()
+
+	types := []interface{}{ecsacs.AckRequest{}}
+	cs := getTestClientServer(mockServer.URL, types, 2)
+	require.NoError(t, cs.Connect())
+	assert.True(t, cs.IsReady(), "expected websocket connection to be ready")
+
+	go func() {
+		messageError <- cs.ConsumeMessages(ctx)
+	}()
+	// Cancel the context.
+	cancel()
+	assert.EqualError(t, <-messageError, "context canceled")
```

Comment:
nit (not blocking, can be added in a followup PR so that this PR can get merged and unblock other work ASAP): 

Can we use `context.Canceled.Error()` here instead of hardcoding "context canceled" (i.e., similar to what we do on L283).